### PR TITLE
service/synapse: Initial setup

### PR DESCRIPTION
**Pull request by e1mo**
Secrets for production not yet added to avoid breaking our setup in case we were to accidentally deploy this.

## Migration plan (proposal)

- [x] Ensure new server (hamilton?) is up and running and being monitored, deploy without synapse (2-n days prior)
- [x] Announce downtime in Allgemein, Große Halle and administration (1 day prior)
- [x] Reduce DNS TTL to 900s (6h to 1 day prior)
- [x] Announce imminent downtime in Allgemein and administration (A few minutes prior to deployment)
- [x] Reduce DNS TTL to 300s
- [x] Stop ~~**and** mask~~ marix-synapse.service service on old host (hopper, `sudo systemctl disable --now matrix-synapse.service; sudo systemctl mask matrix-synapse.service`)
    - Could only disable and not mask the unit (file path foo)
- [x] Shut old server down and take a manual snapshot (just in case)
- [x] Deploy new server with `services.matrix-synapse = lib.mkForce false;` (but with postgres enabled)
- [x] Switch DNS over to new server
- [x] Detach `matrix-media-repo` volume from old server and boot it back up
- [x] Sync database directly to new host: `sudo -u postgres pg_dump --format=custom synapse | ssh hamilton.mon.net.chaos.jetzt "sudo -u postgres pg_restore -d synapse --format=custom --single-transaction --verbose"` (needs SSH with agent forwarding) _Database is to big for dumping and importing separately_
    - `sudo -u postgres pg_dump --format=custom synapse | ssh e1mo@hamilton.mon.net.chaos.jetzt "sudo -u postgres pg_restore -d matrix-synapse --format=custom --single-transaction --verbose --no-owner --role=matrix-synapse"` (Wrong db name on target, and didn't account for different db usernames on target)
- [x] Ensure matrix secrets on new host match old host
- [x] (Only afer db migration is finished) Re-deploy new host with `services.matrix-synapse = true;` and log level of info
- [x] Adjust owner of `synapse_media_store` files (different uids on the servers): `chown matrix-synapse:matrix-synapse /mnt/synapse_media_store -R`
- [x] Monitor logs of matrix synapse for problems
- [x] Validate basic functionality (sending messages, upload files, check federation is working)
- [ ] Ensure all configuration changes are commited to the repo, prolong DNS TTL
- [x] Send end of downtime message to all channels

_Based on [the last migration](https://wiki.chaos.jetzt/telkos/2020-10-18-18-00-ag-administration#ram_probleme_e1mo)_

### CODE REVIEW

**Review by adb-sh — APPROVED** [review]

File: services/matrix/default.nix [hunk 1]
```diff
@@ -0,0 +1,162 @@
+{ lib, config, pkgs, baseDomain, ... }:  let
+  matrixPort = 8008;
+  isDev = (builtins.substring 0 3 baseDomain) == "dev";
+  synapseDb = config.services.matrix-synapse.settings.database.args;
+  initSynapseDb = ''CREATE DATABASE "${synapseDb.database}" WITH OWNER "${synapseDb.user}" ENCODING "UTF8" TEMPLATE template0 LC_COLLATE = "C" LC_CTYPE = "C";'';
```
Comment (e1mo) [inline services/matrix/default.nix:5]: Is encoding UTF8 correct here?

File: services/matrix/default.nix [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        bind_addresses = [ "::1" "127.0.0.1" ];
+        port = matrixPort;
+        type = "http";
+        tls = false;
+        x_forwarded = true;
+        resources = [{
+          names = [ "client" "federation" ];
+          compress = false;
+        }];
+        additional_resources."/_matrix/saml2/pick_username".module = "matrix_synapse_saml_mapper.pick_username_resource";
```
Comment (e1mo) [inline services/matrix/default.nix:148]: Possible dulicate?

File: services/matrix/synapse-log_config.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Based on https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/matrix/synapse-log_config.yaml
+# Just with a log-level of notice
```
Comment (e1mo) [inline services/matrix/synapse-log_config.yaml:2]: Correct that

File: services/matrix/default.nix (hunk 1 above)
Comment (e1mo) [reply services/matrix/default.nix:5]: I took a look at the database on hopper (current prod matrix server, deployed from ansible) and it is currently on C collate:

```
postgres=# select datname, datcollate FROM pg_database WHERE datname = 'synapse';
 datname | datcollate 
---------+------------
 synapse | C
(1 row)
```

The encoding is set to UTF8:

```
postgres=# SELECT pg_encoding_to_char(encoding) FROM pg_database WHERE datname = 'synapse';
 pg_encoding_to_char 
---------------------
 UTF8
```

**Comment by e1mo** [comment]
The build is failing since the secrets (e.g. synapse signing key) for prod are not yet present (to avoid messing up federation and stuff if we were to accidentally deploy).

**Review by supermomme — APPROVED** [review]

File: services/matrix/default.nix (hunk 1 above)
Comment (YtvwlD) [reply services/matrix/default.nix:5]: Interesting.

File: services/matrix/synapse-log_config.yaml (hunk 3 above)
Comment (YtvwlD) [reply services/matrix/synapse-log_config.yaml:2]: This is still notice?

**Review by YtvwlD — APPROVED** [review]

File: services/matrix/synapse-log_config.yaml (hunk 3 above)
Comment (e1mo) [reply services/matrix/synapse-log_config.yaml]: Finally corrected now ;D Thinking that one has changed something is seemingly not the same as actually changing it...

**Comment by e1mo** [comment]
As soon as #26 is merged, I'll re-write this to be for hamilton. However, the migration plan is already at an usable state, so I'd be happy for some feedback regarding it.

**Review by supermomme — APPROVED** [review]
sounds good to me

**Review by adb-sh — APPROVED** [review]
lgtm :D

**Pull request merged**